File: frontend/server/utils/rateLimit.ts
/**
 * IP 限频工具（内存 Map 实现）
 * 同一 IP 每分钟最多 maxRequests 次请求
 */

interface RateLimitEntry {
  count: number
  resetAt: number // 时间戳 ms
}

const ipMap = new Map<string, RateLimitEntry>()

// 每 5 分钟清理过期条目，防止内存泄漏
setInterval(() => {
  const now = Date.now()
  for (const [ip, entry] of ipMap) {
    if (now > entry.resetAt) {
      ipMap.delete(ip)
    }
  }
}, 5 * 60 * 1000)

/**
 * 检查 IP 是否超出限频
 * @param ip 客户端 IP
 * @param maxRequests 每分钟最大请求数（默认 3）
 * @returns true = 允许通过, false = 超出限制
 */
export function checkRateLimit(ip: string, maxRequests = 3): boolean {
  const now = Date.now()
  const entry = ipMap.get(ip)

  if (!entry || now > entry.resetAt) {
    // 无记录或已过期，重新计数
    ipMap.set(ip, { count: 1, resetAt: now + 60 * 1000 })
    return true
  }

  if (entry.count < maxRequests) {
    entry.count++
    return true
  }

  // 超出限制
  return false
}

/**
 * 从请求事件中提取客户端 IP
 */
export function getClientIp(event: any): string {
  // 优先读 x-forwarded-for（Nginx 反代场景）
  const forwarded = getRequestHeader(event, 'x-forwarded-for')
  if (forwarded) {
    return forwarded.split(',')[0].trim()
  }

  // 其次读 x-real-ip
  const realIp = getRequestHeader(event, 'x-real-ip')
  if (realIp) {
    return realIp.trim()
  }

  // 最后用 socket
  return event.node?.req?.socket?.remoteAddress || 'unknown'
}

File: backend/middleware/ratelimit.go
package middleware

import (
	"net/http"
	"strings"
	"sync"
	"time"

	"github.com/gin-gonic/gin"
)

// rateLimitEntry IP 限频记录
type rateLimitEntry struct {
	count   int
	resetAt time.Time
}

var (
	ipMap   = make(map[string]*rateLimitEntry)
	ipMutex sync.Mutex
)

func init() {
	// 每 5 分钟清理过期条目
	go func() {
		ticker := time.NewTicker(5 * time.Minute)
		defer ticker.Stop()
		for range ticker.C {
			ipMutex.Lock()
			now := time.Now()
			for ip, entry := range ipMap {
				if now.After(entry.resetAt) {
					delete(ipMap, ip)
				}
			}
			ipMutex.Unlock()
		}
	}()
}

// checkRateLimit 检查 IP 是否超出限频
func checkRateLimit(ip string, maxRequests int) bool {
	ipMutex.Lock()
	defer ipMutex.Unlock()

	now := time.Now()
	entry, exists := ipMap[ip]

	if !exists || now.After(entry.resetAt) {
		ipMap[ip] = &rateLimitEntry{count: 1, resetAt: now.Add(time.Minute)}
		return true
	}

	if entry.count < maxRequests {
		entry.count++
		return true
	}

	return false
}

// GetClientIP 从请求中提取客户端 IP
func GetClientIP(c *gin.Context) string {
	// 优先读 x-forwarded-for
	if xff := c.GetHeader("X-Forwarded-For"); xff != "" {
		return strings.Split(xff, ",")[0]
	}
	// 其次读 x-real-ip
	if realIP := c.GetHeader("X-Real-IP"); realIP != "" {
		return strings.TrimSpace(realIP)
	}
	return c.ClientIP()
}

// RateLimit 限流中间件
func RateLimit(maxRequests int) gin.HandlerFunc {
	return func(c *gin.Context) {
		ip := GetClientIP(c)
		if !checkRateLimit(ip, maxRequests) {
			c.JSON(http.StatusTooManyRequests, gin.H{
				"error":         true,
				"statusCode":    429,
				"statusMessage": "操作太频繁，请稍后再试",
			})
			c.Abort()
			return
		}
		c.Next()
	}
}
